service/config_provider.go

```diff
@@ -68,50 +68,112 @@ type ConfigProvider interface {
 }
 
 type configProvider struct {
-	locations          []string
-	configMapProviders map[string]config.MapProvider
-	cfgMapConverters   []config.MapConverterFunc
-	configUnmarshaler  configunmarshaler.ConfigUnmarshaler
+	locations           []string
+	configMapProviders  map[string]config.MapProvider
+	configMapConverters []config.MapConverterFunc
+	configUnmarshaler   configunmarshaler.ConfigUnmarshaler
 
 	sync.Mutex
 	closer  config.CloseFunc
 	watcher chan error
 }
 
-// MustNewConfigProvider returns a new ConfigProvider that provides the configuration:
-// * Retrieve the config.Map by merging all retrieved maps from all the config.MapProvider in order.
-// * Then applies all the ConfigMapConverterFunc in the given order.
-// * Then unmarshalls the final config.Config using the given configunmarshaler.ConfigUnmarshaler.
-//
-// The `configMapProviders` is a map of pairs <scheme,Provider>.
+// ConfigProviderOption is an option to change the behavior of ConfigProvider
+// returned by NewConfigProvider()
+type ConfigProviderOption func(opts *configProvider)
+
+// WithConfigMapProvider appends to the default available providers.
+// This provider overwrites any existing provider with the same scheme.
+func WithConfigMapProvider(mp config.MapProvider) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapProviders[mp.Scheme()] = mp
+	}
+}
+
+// WithConfigMapConverters overwrites the default converters.
+func WithConfigMapConverters(c []config.MapConverterFunc) ConfigProviderOption {
+	return func(opts *configProvider) {
+		opts.configMapConverters = c
+	}
+}
+
+// withConfigUnmarshaler overwrites the default unmarshaler.
+// This func is private because providing custom ConfigUnmarshaler is not necessary since users can wrap/implement
+// ConfigProvider if needed to change the resulted config.
```

Comment:
I'm not sure I understand this, or the comment on the same topic in the changelog.  Can you provide an example here of how it would be done?  It seems to me that the `config.Map` has already been turned into a `config.Config` and so there is no ability to influence that translation, only the `config.Map` or the `config.Config`.

This is a capability that exists now.  Are we sure that it is unused and will not present issues if we remove it?  We have made assumptions before about how people use the config processing system and got it very wrong.  I'd like to avoid that happening again.

Reply:
I am 99.99% sure (that the only use-case we heard so far and seen is to alter an unmarshalled `config.Config`, and not to change the unmarshalling logic), that being said, now we have a clear process in place for deprecation. If we make a mistake here, I expect users to come back and tell us that they need that functionality and stop the removal.

I am happy to actually expose this as deprecated, with the "proposed transition", and a note that this will not be removed for at least 2 versions to ensure that possible users are coming and telling us.

Reply:
Reverted this, since is a different change. will submit a different PR to argue about this.